### DISCUSSION TITLE
Check Azure Devops Merge Commit statuses

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#29834] [Suggest an Idea]
### Tell us more.

From a Mend.io customer who's using Renovate:

```
I believe it's how Renovate checks the statuses and how Azure DevOps PR work when you have build policy set with build expiration.

In short, Azure DevOps in such configuration creates new temporary commit for the purpose of building PR's commit. Then that temporary commit is tagged with the build status while the original commit (latest commit on PR's branch) is left un-tagged.

From Azure DevOps API perspective:
GET https://dev.azure.com/{{organization}}/_apis/git/pullrequests/{{pullRequestId}}?api-version=4.1
returns object that contains:

`lastMergeSourceCommit.commitId` - it is the latest commit on PR's branch

`lastMergeCommit.commitId` - it is that temporary commit created for PR's build

If you then use the
GET https://dev.azure.com/{{organization}}/{{project}}/_apis/git/repositories/{{repositoryId}}/commits/{{commitId}}/statuses?latestOnly=true&api-version=4.1
with above commit IDs you can tell that only temporary commit contains the build status, while the latest commit on PR's branch has no build status at all.
```

First point of clarification: Renovate uses the AZDO SDK instead of API directly.

Renovate's `getBranchStatus()` implementation for AZDO is here: https://github.com/renovatebot/renovate/blob/a581f7b720952a3258e0e8dfb9348a5f947ba69b/lib/modules/platform/azure/index.ts#L405-L446

It relies on this function: https://github.com/renovatebot/renovate/blob/a581f7b720952a3258e0e8dfb9348a5f947ba69b/lib/modules/platform/azure/index.ts#L361-L379

Like with other platforms it:
- Gets the branch: https://learn.microsoft.com/en-us/javascript/api/azure-devops-extension-api/gitrestclient#azure-devops-extension-api-gitrestclient-getbranch
- Extracts the commit.commitId from the result: https://learn.microsoft.com/en-us/javascript/api/azure-devops-extension-api/gitbranchstats
- Gets statuses for the commit ID of the branch: https://learn.microsoft.com/en-us/javascript/api/azure-devops-extension-api/gitrestclient#azure-devops-extension-api-gitrestclient-getstatuses

Theoretically instead:
- `getPullRequest`: https://learn.microsoft.com/en-us/javascript/api/azure-devops-extension-api/gitrestclient#azure-devops-extension-api-gitrestclient-getstatuses
- get [lastMergeCommit](https://learn.microsoft.com/en-us/javascript/api/azure-devops-extension-api/gitpullrequest#azure-devops-extension-api-gitpullrequest-lastmergecommit) from [the results](https://learn.microsoft.com/en-us/javascript/api/azure-devops-extension-api/gitpullrequest)

Proposal:
- If no PR then we use the existing approach
- If there's a PR and no `lastMergeCommit` or it's the same as `lastMergeSourceCommit` then use that commit (should in theory be the same result as today)
- If there's a PR and `lastMergeCommit` exists then use it exclusively, or check both it and `lastMergeSourceCommit` and combine results?

Question:
- Onto which commit should Renovate place its internal status checks?

**Comment by davidw-verbund** [comment]
It would also be possible to check the policy evaluation of the PR with this API call https://learn.microsoft.com/en-us/rest/api/azure/devops/policy/evaluations/list?view=azure-devops-rest-7.2
